include/openPMD/auxiliary/Logging.hpp

```diff
@@ -0,0 +1,93 @@
+/* Copyright 2019 Fabian Koller
+ *
+ * This file is part of openPMD-api.
+ *
+ * openPMD-api is free software: you can redistribute it and/or modify
+ * it under the terms of of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * openPMD-api is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with openPMD-api.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#if openPMD_HAVE_LOGGING
+#   include <spdlog/spdlog.h>
+#endif
+
+namespace openPMD
+{
+namespace auxiliary
+{
+namespace log
+{
+
+enum class Level
+{
+    TRACE,
```

Comment:
can we also add a "timing" mode in this or another level? "throughput" (bytes/s) would also be awesome.
cc @franzpoeschel

Reply:
Interesting idea. That would mean we count bytes during all backend IO and timestamps in the fronend calls.

Intuitively, this would result in a different code structure.

```cpp
auto series = openPMD::Series("data%T.h5", openPMD::AccessType::CREATE);
MeshRecordComponent rho = series.iterations[1].meshes["rho"][MeshRecordComponent::SCALAR];

uint64_t patch_size = ...
auto patch = std::make_shared< double >(patch_size);

// assume DEFER
rho.store_chunk(patch, {0}, {patch_size});
LOG_TIME(series.flush());     /* example.cpp:9 series.flush() took 0.314s */

std::array< std::shared_prt< double >, 42 > lots_of_patches = ...
for( auto patch : lots_of_patches )
    rho.store_chunk(patch, ...);
LOG_THROUGHPUT(series.flush());  /* example.cpp:14 series.flush() averaged 314MB/s */
```

I'll see what I can do. This will be pretty invasive, though.

Reply:
Doesn't need to go into the first log version, we can add those in a follow-up if you like.